src/lib/krb5/krb/get_etype_salt_s2kp.c

```diff
@@ -0,0 +1,118 @@
+ * to distribution of the software without specific, written prior
+ * permission.  Furthermore if you modify this software you must label
+ * your software as modified software and not distribute it in such a
+ * fashion that it might be confused with the original M.I.T. software.
+ * M.I.T. makes no representations about the suitability of
+ * this software for any purpose.  It is provided "as is" without express
+ * or implied warranty.
+ */
+
+#include "k5-int.h"
+#include "init_creds_ctx.h"
+
+/*
+ * Send an AS_REQ to KDC and return etype, salt and s2kparams from AS_REP/ERROR
+ */
+krb5_error_code KRB5_CALLCONV
+krb5_get_etype_salt_s2kp(krb5_context context, krb5_principal principal,
+                         krb5_enctype *enctypes, krb5_enctype *enctype,
+                         krb5_data *salt, krb5_data *s2kparams)
+{
+    krb5_get_init_creds_opt *opt;
+    krb5_init_creds_context icc = NULL;
+    krb5_data reply = empty_data(), request = empty_data(), realm = empty_data();
+    krb5_error *error;
+    krb5_kdc_rep *asrep;
+    krb5_pa_data **padata;
+    unsigned int flags;
+    int master = 0, nenctypes = 0;
+    krb5_error_code retval = 0;
+
+    retval = krb5_get_init_creds_opt_alloc(context, &opt);
+    if (retval)
+        goto cleanup;
+    if (enctypes) {
+        nenctypes = k5_count_etypes(enctypes);
+        if (nenctypes)
+            krb5_get_init_creds_opt_set_etype_list(opt, enctypes, nenctypes);
+    }
+
+    retval = krb5_init_creds_init(context, principal, NULL, NULL, 0, opt, &icc);
+    if (retval)
+        goto cleanup;
+    retval = krb5_init_creds_step(context, icc, &reply, &request, &realm,
+                                  &flags);
+    if (retval)
+        goto cleanup;
+    if (flags != KRB5_INIT_CREDS_STEP_FLAG_CONTINUE) {
+        retval = KRB5KRB_AP_ERR_MSG_TYPE;
+        goto cleanup;
+    }
+    retval = krb5_sendto_kdc(context, &request, &realm, &reply, &master, 1);
+    if (retval)
+        goto cleanup;
+
+    if (krb5_is_krb_error(&reply)) {
+        retval = decode_krb5_error(&reply, &error);
+        if (retval)
+            goto cleanup;
+        retval = decode_krb5_padata_sequence(&error->e_data, &padata);
+        if (retval)
+            goto cleanup;
+        retval = k5_get_etype_info(context, icc, padata);
+        if (retval) {
+            krb5_free_error(context, error);
+            goto cleanup;
+        }
+        krb5_free_pa_data(context, padata);
+        krb5_free_error(context, error);
+    } else if (krb5_is_as_rep(&reply)) {
+        retval = decode_krb5_as_rep(&reply, &asrep); 
+        if (retval)
+            goto cleanup;
+        retval = k5_get_etype_info(context, icc, asrep->padata);
+        if (retval) {
+            krb5_free_kdc_rep(context, asrep);
+            goto cleanup;
+        }
+        krb5_free_kdc_rep(context, asrep);
+    } else
+        return KRB5KRB_AP_ERR_MSG_TYPE;
+    *enctype = icc->etype;
```

Comment:
It's possible for the KDC reply not to include any etype-info or salt information.  In this case, k5_get_etype_info() will still return 0 (it will fall back to get_salt() which will return 0 after not finding PA-PW-SALT or PA-AFS3-SALT elements), but icc->etype won't have been set.  This can happen if the principal has no keys, for instance, and we need to handle it gracefully rather than return the wrong result.

The best way I can think of to detect this case is to set icc->etype to ENCTYPE_NULL before calling k5_get_etype_info(), and then check if it's still ENCTYPE_NULL afterwards.  (krb5_init_creds_init() sets icc->etype to the first request enctype, so we need the assignment.)